File: libraries/clipper/clipper/polyclipping/tags/ver 3.0/cairo/CPP/cairo_clipper.cpp
/*******************************************************************************
*                                                                              *
* Author    :  Angus Johnson                                                   *
* Version   :  0.6                                                             *
* Date      :  6 December 2010                                                 *
* Copyright :  Angus Johnson                                                   *
*                                                                              *
* License:                                                                     *
* Use, modification & distribution is subject to Boost Software License Ver 1. *
* http://www.boost.org/LICENSE_1_0.txt                                         *
*                                                                              *
* Modified by Mike Owens to support coordinate transformation and tolerance    *
*******************************************************************************/

#include <cairo/cairo.h>
#include "clipper.hpp"
#include "cairo_clipper.hpp"

namespace clipper {
  namespace cairo {

    namespace {
      inline void transform_point(cairo_t* pen,
                                  Transform transform,
                                  double* x, double* y)
      {
        switch (transform) 
        {
          case tDeviceToUser:
            cairo_device_to_user(pen, x, y);
            break;
          case tUserToDevice:
            cairo_user_to_device(pen, x, y);
            break;
          default:
            ;
        }
      }
    }

    void cairo_to_clipper(cairo_t* cr,
                          TPolyPolygon &pg,
                          Transform transform,
                          double tolerance)
    {
      pg.clear();

      /* Avoid a push/pop if not required. */
      cairo_path_t *path = 0;
      if (tolerance != 0.0) {
        cairo_save(cr);
        cairo_set_tolerance(cr, tolerance);
        path = cairo_copy_path_flat(cr);
        cairo_restore(cr);
      } else {
        path = cairo_copy_path_flat(cr);
      }

      int poly_count = 0;
      for (int i = 0; i < path->num_data; i += path->data[i].header.length) {
        if( path->data[i].header.type == CAIRO_PATH_CLOSE_PATH) {
          poly_count++;
        }
      }

      pg.resize(poly_count);
      int i = 0, pc = 0;
      while (pc < poly_count)
      {
        int vert_count = 1;
        int j = i;
        while(j < path->num_data &&
          path->data[j].header.type != CAIRO_PATH_CLOSE_PATH)
        {
          if (path->data[j].header.type == CAIRO_PATH_LINE_TO)
            vert_count++;
          j += path->data[j].header.length;
        }
        pg[pc].resize(vert_count);
        if (path->data[i].header.type != CAIRO_PATH_MOVE_TO) {
          pg.resize(pc);
          break;
        }
        pg[pc][0].X = path->data[i+1].point.x;
        pg[pc][0].Y = path->data[i+1].point.y;
        transform_point(cr, transform, &pg[pc][0].X, &pg[pc][0].Y);

        i += path->data[i].header.length;

        j = 1;
        while (j < vert_count && i < path->num_data &&
          path->data[i].header.type == CAIRO_PATH_LINE_TO) {
          pg[pc][j].X = path->data[i+1].point.x;
          pg[pc][j].Y = path->data[i+1].point.y;
          transform_point(cr, transform, &pg[pc][j].X, &pg[pc][j].Y);
          j++;
          i += path->data[i].header.length;
        }
        pc++;
        i += path->data[i].header.length;
      }
      cairo_path_destroy(path);
    }
    //--------------------------------------------------------------------------

    void clipper_to_cairo(const TPolyPolygon &pg,
                          cairo_t* cr,
                          Transform transform)
    {
      for (size_t i = 0; i < pg.size(); ++i)
      {
        size_t sz = pg[i].size();
        if (sz < 3)
          continue;
        cairo_new_sub_path(cr);
        for (size_t j = 0; j < sz; ++j) {
          double x = pg[i][j].X,
                 y = pg[i][j].Y;
          transform_point(cr, transform, &x, &y);
          cairo_line_to(cr, x, y);
        }
        cairo_close_path(cr);
      }
    }
    //--------------------------------------------------------------------------

  }
}

File: libraries/clipper/clipper/polyclipping/tags/ver 3.0/cairo/Delphi/cairo_clipper.pas
unit cairo_clipper;

(*******************************************************************************
*                                                                              *
* Author    :  Angus Johnson                                                   *
* Version   :  0.50                                                            *
* Date      :  30 October 2010                                                 *
* Website   :  http://www.angusj.com                                           *
* Copyright :  Angus Johnson 2010                                              *
*                                                                              *
* License:                                                                     *
* Use, modification & distribution is subject to Boost Software License Ver 1. *
* http://www.boost.org/LICENSE_1_0.txt                                         *
*                                                                              *
*******************************************************************************)

interface

uses
  SysUtils, Classes, Cairo, clipper;

function DoublePointArrayToCairo(const fpa: TArrayOfArrayOfDoublePoint;
  cairo: Pcairo_t): boolean;
function CairoToDoublePointArray(cairo: Pcairo_t;
  out fpa: TArrayOfArrayOfDoublePoint): boolean;

implementation

type
  PCairoPathDataArray = ^TCairoPathDataArray;
  TCairoPathDataArray =
    array [0.. MAXINT div sizeof(cairo_path_data_t) -1] of cairo_path_data_t;

function DoublePointArrayToCairo(const fpa: TArrayOfArrayOfDoublePoint;
  cairo: Pcairo_t): boolean;
var
  i,j: integer;
begin
  result := assigned(cairo);
  if not result then exit;
  for i := 0 to high(fpa) do
  begin
    cairo_new_sub_path(cairo);
    for j := 0 to high(fpa[i]) do
      with fpa[i][j] do
        cairo_line_to(cairo,X,Y); //does cairo_move_to() on first iteration
    cairo_close_path(cairo);
  end;
end;
//------------------------------------------------------------------------------

function CairoToDoublePointArray(cairo: Pcairo_t;
  out fpa: TArrayOfArrayOfDoublePoint): boolean;
const
  buffLen1: integer = 32;
  buffLen2: integer = 128;
var
  i,currLen1, currLen2: integer;
  pdHdr: cairo_path_data_t;
  path: Pcairo_path_t;
  currPos: TDoublePoint;
begin
  result := false;
  setlength(fpa, buffLen1);
  currLen1 := 1;
  currLen2 := 0;
  currPos := DoublePoint(0,0);
  i := 0;
  path := cairo_copy_path_flat(cairo);
  try
    while i < path.num_data do
    begin
      pdHdr := PCairoPathDataArray(path.data)[i];
      case pdHdr.header._type of
      CAIRO_PATH_CLOSE_PATH:
        begin
          if currLen2 > 1 then //ie: ignore if < 3 points (not a polygon)
          begin
            setlength(fpa[currLen1-1], currLen2);
            setlength(fpa[currLen1], buffLen2);
            inc(currLen1);
          end;
          currLen2 := 0;
          currPos := DoublePoint(0,0);
          result := true;
        end;
      CAIRO_PATH_MOVE_TO, CAIRO_PATH_LINE_TO:
        begin
          result := false;
          if (pdHdr.header._type = CAIRO_PATH_MOVE_TO) and
            (currLen2 > 0) then break; //ie enforce ClosePath for polygons
          if (currLen2 mod buffLen2 = 0) then
            SetLength(fpa[currLen1-1], currLen2 + buffLen2);
          with PCairoPathDataArray(path.data)[i+1].point do
            currPos := DoublePoint(x,y);
          fpa[currLen1-1][currLen2] := currPos;
          inc(currLen2);
        end;
      end;
      inc(i, pdHdr.header.length);
    end;
  finally
    cairo_path_destroy(path);
  end;
  dec(currLen1); //ie enforces a ClosePath
  setlength(fpa, currLen1);
end;
//------------------------------------------------------------------------------

end.
